tests/test_lock.py

```diff
@@ -49,12 +57,95 @@ def test_lock_file_create_fails_retried(self, context):
 
         self.assertTrue(
             os.path.exists(
-                os.path.join(lck.BASE_DIR, 'namespace', 'somename')))
+                os.path.join(lock.Lock.BASE_DIR, 'namespace', 'somename')))
+
+    def setup_fcntl_return(self, context):
+        context.mock_fcntl = mock.MagicMock(spec=fcntl.fcntl)
+        context.mock_fcntl.return_value = struct.pack(
+            flock.Flock.FORMAT, fcntl.F_WRLCK, 0, 0, 0, 0)
+
+    @testlib.with_context
+    def test_lock_acquire_release(self, context):
+        self.setup_fcntl_return(context)
+
+        lck = lock.Lock("somename")
+
+        lck.acquire()
+
+        self.assertTrue(lck.held())
+
+        lck.release()
+
+        self.assertFalse(lck.held())
+
+    @testlib.with_context
+    def test_lock_acquire_noblock_release(self, context):
+        self.setup_fcntl_return(context)
+
+        lck = lock.Lock("somename")
+
+        lck.acquireNoblock()
+
+        self.assertTrue(lck.held())
+
+        lck.release()
+
+        self.assertFalse(lck.held())
+
+    @testlib.with_context
+    def test_lock_acquire_twice_release(self, context):
+        self.setup_fcntl_return(context)
+
+        lck1 = lock.Lock("somename")
+
+        lck1.acquire()
+
+        self.assertTrue(lck1.held())
+
+        # This should be the same lock as lck1
+        lck2 = lock.Lock("somename")
+        lck2.acquire()
+
+        self.assertTrue(lck2.held())
+        lck2.release()
+
+        # As lck1 and lck2 refer to the same lock they should still be held
+        self.assertTrue(lck2.held())
+        self.assertTrue(lck1.held())
+
+        lck1.release()
+
+        self.assertFalse(lck1.held())
+
+    @testlib.with_context
+    def test_lock_acquire_twice_noblock_release(self, context):
+        self.setup_fcntl_return(context)
+
+        lck1 = lock.Lock("somename")
+
+        lck1.acquire()
+
+        self.assertTrue(lck1.held())
```

Comment:
Perhaps some value in copy pasting these tests twice more so that we get all four permutations.
aquire,aquire DONE
aquire, noblock DONE
noblock, aquire MISSING
noblock, noblock MISSING

Reply:
Done